Code/GraphMol/MarvinParse/MarvinDefs.cpp

```diff
@@ -0,0 +1,2978 @@
+  }
+
+  // find the atoms to be deleted
+
+  std::vector<MarvinAtom *> atomsToDelete;
+  std::vector<MarvinBond *> bondsToDelete;
+  std::vector<MarvinBond *> orphanedBonds;
+  std::vector<MarvinMolBase *> sgroupsToDelete;
+
+  for (MarvinAtom *atomPtr : this->atoms) {
+    // Atoms in the group but NOT in the parent part are to be deleted
+    if (!boost::algorithm::contains(this->parentAtoms,
+                                    std::vector<std::string>{atomPtr->id},
+                                    atomRefInAtoms)) {
+      atomsToDelete.push_back(atomPtr);
+    }
+  }
+
+  for (MarvinBond *bondPtr : actualParent->bonds) {
+    bool atom1ToBeDeleted = boost::algorithm::contains(
+        atomsToDelete, std::vector<std::string>{bondPtr->atomRefs2[0]},
+        atomRefInAtoms);
+    bool atom2ToBeDeleted = boost::algorithm::contains(
+        atomsToDelete, std::vector<std::string>{bondPtr->atomRefs2[1]},
+        atomRefInAtoms);
+    if (atom1ToBeDeleted && atom2ToBeDeleted) {
+      bondsToDelete.push_back(bondPtr);
+    } else if (atom1ToBeDeleted) {
+      orphanedBonds.push_back(bondPtr);
+      swap(bondPtr->atomRefs2[0],
+           bondPtr->atomRefs2[1]);  // make sure the first ref atom is the
+                                    // orphaned one (still in the mol), and
+                                    // the second is the atom to be deleted
+      bondsToDelete.push_back(
+          bondPtr);  // one of each pair of orphaned bonds will be UNdeleted
+    } else if (atom2ToBeDeleted) {
+      orphanedBonds.push_back(bondPtr);
+      bondsToDelete.push_back(
+          bondPtr);  // one of each pair of orphaned bonds will be UNdeleted
+    }
+  }
+
+  // there must be zero two or four orphaned bonds
+  // zero happens when the whole group is replicated and not chained
+  // two is most common - the replicates being deleted are connected to each
+  // other 4 happens when the parent replicate (not deleted) is in between two
+  // replicates that are deleted.
+  //  there are two orphaned bonds on the parent replicate, and two on the
+  //  rest of the molecule
+
+  if (orphanedBonds.size() != 0 && orphanedBonds.size() != 2 &&
+      orphanedBonds.size() != 4) {
+    throw FileParseException(
+        "Error: there should be zero, two or four orphaned bonds while contracting a MultipleSgroup");
+  }
+
+  // delete any of this group's children that were using the atoms to be
+  // deleted
+
+  for (auto childSgroup : sgroups) {
+    if (childSgroup->isSgroupInSetOfAtoms(atomsToDelete) !=
+        SgroupNotInAtomSet) {
+      sgroupsToDelete.push_back(childSgroup);
+    }
+  }
+
+  // now fix the orphaned bonds - The first gets the atoms from the matched
+  // second orphan bond  and was NOT removed. the matched second orphaned bond
+  // is deleted
+
+  while (orphanedBonds.size() > 0) {
+    int matchedOrphanBondIndex = (-1);
+    auto orphanedBondToFix = orphanedBonds[0];
+    orphanedBonds.erase(orphanedBonds.begin());
```

Comment:
calling `erase` on a vector, which you do repeatedly in this function, is pretty ineffiicent. It might be worth exploring whether or not another approach makes sense. (Or it might not matter... this piece of code may not be called that often)

Reply:
the orphanedBonds vector never has more than 4 elements, so it is probably not useful to work too hard at this.  The order doesn't matter, so we could just copy the last one into the one to be removed, and truncate the vector.  This is almost certainly not worth it.  It seems to be rare that there are 4 elements - usually 2 or 0.